lib/http_aws_sigv4.c

```diff
@@ -66,6 +68,125 @@ static void sha256_to_hex(char *dst, unsigned char *sha, size_t dst_l)
   }
 }
 
+
+
+static CURLcode make_headers(struct Curl_easy *data,
+                             const char *hostname,
+                             const char *timestamp,
+                             const char *provider1_low,
+                             const char *provider1_mid,
+                             char **canonical_headers,
+                             char **signed_headers)
+{
+  char *date_hdr_key =  curl_maprintf("X-%s-Date", provider1_mid);
+  char *date_full_hdr = curl_maprintf("x-%s-date:%s",
+                                      provider1_low, timestamp);
+#ifdef DEBUGBUILD
+  char *force_host = getenv("CURL_FORCETIME");
+#endif
+  struct curl_slist *head;
+  int ret = CURLE_OUT_OF_MEMORY;
+  struct curl_slist *l;
+  int again = 1;
+
+  if(Curl_checkheaders(data, "Host")) {
+    head = NULL;
+  }
+#ifdef DEBUGBUILD
+  else if(force_host) {
+    head = curl_slist_append(NULL, "host:127.0.0.1");
+  }
+#endif
+  else {
+    char *full_host = data->state.aptr.host ? data->state.aptr.host :
+      curl_maprintf("host:%s", hostname);
```

Comment:
No space after the colon? And what if `curl_maprintf()` returns NULL?

In what situations is `data->state.aptr.host` NULL when this is called? What then if there's a custom port number or an IPv6 address host name?

Reply:
TBH I really don't know if not having data->state.aptr.host is possible, I've made this ternary out of fear of having full_host NULL, which would be a bug.
Should I remove the condition ?

Reply:
It is probably wise to keep it.

Reply:
No the CanonicalHeaders must contain `Lowercase(HeaderName) + ':' + Trimall(HeaderValue) + '\n'`.
so no space before and after the Value.
I don't see why it wouldn't work with the current code with IPv6 ?
I guess I could try to add manually the port, if the ip is set using `conn->host.name`, though most case where a port is needed should be handle using `data->state.aptr.host`. 
My guess is that this would be too much code for handling a too niche corner case, maybe add a comment saying we don't handle port if host is set using `conn->host.name` is enough ?

Reply:
Please note that `data->state.aptr.host` consists of `\r\n` on its end. You have to trim them in order not to create extra empty lines in CanonicalHeaders.